pymedphys/_dicom/dose.py

```diff
@@ -179,58 +179,81 @@ def profile(displacements, depth, direction, dose_dataset, plan_dataset):
     return extracted_dose
 
 
-def _get_indices(z_list, z_val):
-    indices = np.array([item[0] for item in z_list])
-    # This will error if more than one contour exists on a given slice
-    desired_indices = np.where(indices == z_val)[0]
-    # Multiple contour sets per slice not yet implemented
+def get_dose_grid_structure_mask(structure_name, structure_dataset, dose_dataset):
+    x_dose, y_dose, z_dose = xyz_axes_from_dataset(dose_dataset)
 
-    return desired_indices
+    xx, yy = np.meshgrid(x_dose, y_dose)
+    points = np.swapaxes(np.vstack([xx.ravel(), yy.ravel()]), 0, 1)
 
+    x_structure, y_structure, z_structure = pull_structure(
+        structure_name, structure_dataset
+    )
 
-def get_dose_grid_structure_mask(structure_name, dcm_struct, dcm_dose):
-    x_dose, y_dose, z_dose = xyz_axes_from_dataset(dcm_dose)
-
-    xx_dose, yy_dose = np.meshgrid(x_dose, y_dose)
-    points = np.swapaxes(np.vstack([xx_dose.ravel(), yy_dose.ravel()]), 0, 1)
+    structure_z_values = []
+    for item in z_structure:
+        item = np.unique(item)
+        if len(item) != 1:
+            raise ValueError("Only one z value per contour supported")
+        structure_z_values.append(item[0])
+
+    structure_z_values = np.sort(structure_z_values)
+    unique_structure_z_values = np.unique(structure_z_values)
+
+    if np.any(structure_z_values != unique_structure_z_values):
+        raise ValueError("Only one contour per slice is currently supported")
+
+    sorted_dose_z = np.sort(z_dose)
+
+    first_dose_index = np.where(sorted_dose_z == structure_z_values[0])[0][0]
+    for i, z_val in enumerate(structure_z_values):
+        dose_index = first_dose_index + i
+        if structure_z_values[i] != sorted_dose_z[dose_index]:
+            raise ValueError(
+                "Only contours where both, there are no gaps in the "
+                "z-axis of the contours, and the contour axis and dose "
+                "axis, are aligned are supported."
+            )
 
-    x_structure, y_structure, z_structure = pull_structure(structure_name, dcm_struct)
-    structure_z_values = np.array([item[0] for item in z_structure])
+    mask_yxz = np.zeros((len(y_dose), len(x_dose), len(z_dose)), dtype=bool)
 
-    mask = np.zeros((len(y_dose), len(x_dose), len(z_dose)), dtype=bool)
+    for structure_index, z_val in enumerate(structure_z_values):
+        dose_index = int(np.where(z_dose == z_val)[0])
 
-    for z_val in structure_z_values:
-        structure_indices = _get_indices(z_structure, z_val)
+        if z_structure[structure_index][0] != z_dose[dose_index]:
+            raise ValueError("Expected structure and dose indices to align")
```

Comment:
I think it is clearer if you state what the error is, explicitly.
"Structure and dose indices do not align"

but this isn't a reason to hold up the PR